### ISSUE TITLE
react-native compatibility

### ISSUE THREAD

**Issue by davidpfahler** [pouchdb/pouchdb#3787]
I spent the last two days figuring out how to make PouchDB compatible with react-native. To make it work I had to work around a couple of things that I would like to address below. This is intended as a comprehensive list for discussion on how we could make this easier, e.g. by eliminating the need for workarounds. If you just need to get it working, skip to the TLDR at the very bottom.

## webpack
Issue #3319 is concerned with getting PouchDB working with webpack. It is to be expected that close to 100% of react-native users are going to also use webpack. react-native brings additional challenges to getting PouchDB working with webpack. As far as I understand, react-native requires some files ("native components", see https://github.com/roman01la/react-native-babel/tree/dep-graph#react-native-in-es6-with-webpack-and-babel for details) via the packager that should not be resolved by webpack beforehand. To avoid this, we need to define them as webpack externals. The most elegant way to do this, that I found, is to use the `DependencyGraph` that comes with react-native. 

This is relevant for PouchDB, because - get this - one of the dependencies spit out by DependencyGraph is called `merge`. Now, the way `roman01la/react-native-babel` currently handles externals is using a RegEx. If you employ this tactic, you match everything with `merge` in a require call. As you might now, PouchDB has quite some files called `merge.js` and is requiring them. To avoid webpack failure, you need to make sure that you specify the webpack externals as exact matches, as you can see in my [pull request](https://github.com/davidpfahler/react-native-babel/commit/d6c92ab0f65461e6f3c82b915a1fc62fe9e6873c).

I found this config to work for me: [`webpack.config.js`](https://github.com/davidpfahler/react-native-babel-example/blob/47265eb8b55c0a99e04fddfbea45f3c4e48aaeea/webpack.config.js), [`webpack-watch.js`](https://github.com/davidpfahler/react-native-babel-example/blob/c6129ef7bee58a5bf7b9dc9e623ea9be00c037ff/webpack-watch.js).

The plugins in the config are probably optional. What is important is that you have specified the loaders. Also make sure you `npm install`ed the loaders, in this case, `npm install -D babel-loader json-loader`. If the warning about `lie.js` bothers you, add `noParse: /lie.js/` to your config.
The `webpack-watch.js` not only provides watch functionality, but also handles the issue with the `DependencyGraph`.

## package.json of PouchDB
The above config will still fail. When webpack parses `leveldown`, it logs the following warnings and error:

```
WARNING in ./~/pouchdb/~/leveldown/~/bindings/bindings.js
Critical dependencies:
76:22-40 the request of a dependency is an expression
76:43-53 the request of a dependency is an expression
 @ ./~/pouchdb/~/leveldown/~/bindings/bindings.js 76:22-40 76:43-53

WARNING in ./~/pouchdb/~/leveldown/~/bindings/README.md
Module parse failed: /Users/djpfahler/localhost/react-native-babel-example/node_modules/pouchdb/node_modules/leveldown/node_modules/bindings/README.md Line 2: Unexpected token ===
You may need an appropriate loader to handle this file type.
| node-bindings
| =============
| ### Helper module for loading your native module's .node file
| 
 @ ./~/pouchdb/~/leveldown/~/bindings ^\.\/.*$

ERROR in ./~/pouchdb/~/leveldown/~/bindings/bindings.js
Module not found: Error: Cannot resolve module 'fs' in /Users/djpfahler/localhost/react-native-babel-example/node_modules/pouchdb/node_modules/leveldown/node_modules/bindings
 @ ./~/pouchdb/~/leveldown/~/bindings/bindings.js 6:9-22
```

I do not fully understand why this is the case. I will keep digging and try to find a good solution, but for now, I just added `leveldown: false` to the `browser` field in the `package.json` of PouchDB (see [my fork](https://github.com/davidpfahler/pouchdb/commit/cc871f4220e16c429c9fcf051cf434571687d01d)). This prevents webpack to parse leveldown in its standard configuration. This might be altered by the `target` setting, which I'm about to investigate (updates will follow).

## asyncstorage-down custom adapter
Now that webpack is building our bundle and we can `require('pouchdb')` in our react-native app, we need some adapter that works for react-native. In react-native, the only available persistent storage is [`AsyncStorage`](http://facebook.github.io/react-native/docs/asyncstorage.html#content). GitHub user tradle was kind enough to create a leveldown layer on top of that called [`asyncstorage-down`](https://github.com/tradle/asyncstorage-down). I had to update it for the latest react-native release, because otherwise (at least) debugging (maybe more) didn't work, see [my PR](https://github.com/tradle/asyncstorage-down/pull/1).

Unfortunately, we cannot simply use the `leveldb` adapter with `asyncstorage-down`. If we could, we'd have the problem that it's not added to PouchDB when `process.browser` is `true`. So we need a custom adapter. nolanlawson was kind enough to send me his advice and work on such an adapter, see [nolanlawson/pouchdb-async-storage](https://github.com/nolanlawson/pouchdb-async-storage). Based on that, I developed the following code inside [my react-native example app](https://github.com/davidpfahler/react-native-babel-example/tree/pouchdb), which adds and uses a custom adapter:

* [`leveldb-alt.jsx`](https://github.com/davidpfahler/react-native-babel-example/blob/pouchdb/src/lib/leveldb-alt.jsx) implements the adapter for `asyncstorage-down`. I copied this code from nolan and have to admit I don't fully understand it.
* [`db.jsx`](https://github.com/davidpfahler/react-native-babel-example/blob/pouchdb/src/lib/db.jsx) requires both `pouchdb` and the custom adapter and ads the adapter to PouchDB using a small config.
* [`app.jsx` imports `db.jsx`](https://github.com/davidpfahler/react-native-babel-example/blob/pouchdb/src/app.jsx#L5)

Please note that I have to unshift the custom adapter to the front of the preferredAdapters array, because otherwise idb is chosen, although IndexedDB is not present.

## summary or workaround
:tada: We have a PouchDB running and working inside react-native. This is cool, but was too complicated. Here is a summary of the workarounds used (enumerated to be addressable, not because of importance):

1. changed `package.json` of PouchDB
2. added custom adapter to the app
3. use a webpack config with properly defined externals

I think the biggest issues I came across was the "discrimination" of leveldown adapters when process.browser is true. If I understand correctly, leveldown is just a low level api/interface that pouchdb should be able to use no matter what the environment or what's behind that interface.

## TLDR
Please check out my [react-native-babel-example with pouchdb](https://github.com/davidpfahler/react-native-babel-example/tree/9c537350256bd64d3c1ba2e5a7df5eb7ec8ca149). If contains my webpack config and a customized pouchdb (only package.json modified!) that will get you up and running. The `src/lib` folder contains the custom adapter and database. You should be able to get going immediately.

**Comment by nolanlawson** [comment]
For everybody else: @davidpfahler and I had a long chat over IRC and think we've resolved most of these issues. PouchDB in React Native is happening!!

**Comment by svnlto** [comment]
:clap:

**Comment by robertjpayne** [comment]
@nolanlawson @davidpfahler I'd love to help out here, one thing is React Native's AsyncStorage is very very unsuitable for any heavy access or large amounts of key/values.

**Comment by davidpfahler** [comment]
@robertjpayne Thanks, I think we should definitely have a discussion as to what would be suitable as a data store for react-native + pouch. However, we first have to get it running at all. I was spending a couple of days on it and made some progress which landing in the [for-react-native](https://github.com/pouchdb/pouchdb/tree/for-react-native) branch. However, I'm now stuck with no errors and pouchdb still silently failing the replication. It's very likely to be something with react-native's XMLHttpRequest implementation.

**Comment by robertjpayne** [comment]
@davidpfahler ah I was under the impression you had it working already -- you need some help tracking it down?

**Comment by davidpfahler** [comment]
@robertjpayne Right now, I lack the time to do a deep dive into all the code to see what could be the problem. If you like to help, that would be very kind. Please find my [repo's pouchdb branch here](https://github.com/davidpfahler/react-native-babel-example/tree/pouchdb). Sorry for the mess it is currently.

**Comment by nolanlawson** [comment]
@robertjpayne `asyncstorage-down` is also a pretty inefficient usage even of AsyncStorage, because it's based on `localstorage-down`, which has no key iteration and thus keeps all keys in memory. :)

I think the point of this is just to get a quick proof-of-concept up and running, and plus it'll work fine for small databases that need to be replicated.

Also FWIW most of the issues @davidpfahler and I have been running into so far have been related to Webpack and JavaScriptCore, not AsyncStorage.

**Comment by robertjpayne** [comment]
@nolanlawson @davidpfahler cool, I'll spend some time today and see what I can get working.

**Comment by robertjpayne** [comment]
@nolanlawson @davidpfahler I'm getting closer -- correct me if I'm wrong but it appears the 'necessary' fixes for react-native have been merged into master?

I've got regular PouchDB (not the for-react-native) branch running just fine using https://github.com/mjohnston/react-native-webpack-server.

I've added some error watchers on the feed replication and it's giving an error saying the adapter cannot be found so I'm attempting to track that down first.

**Comment by robertjpayne** [comment]
Ok got the adapter working -- I'll look into what's failing now.

**Comment by nolanlawson** [comment]
Yes, I think those "necessary" fixes been released in 3.5.0 even.

**Comment by robertjpayne** [comment]
@nolanlawson yea they are fixed, but I found the async storage adapter repo has a reference to the for-react-native branch so it bugs it up.

Anyways it definitely appears to be something to do with XMLHttpRequest I'm just getting back into it now

**Comment by robertjpayne** [comment]
@nolanlawson @davidpfahler 

Getting closer but it's very hard to get a stack trace out of react-native that is worth anything. The replication fails after getting a 404 on something like:

`http://localhost:3000/db/exampledb/_local/gp_4uQVfYGK_B2JCuzudIg%3D%3D?_nonce=1431416721250 => 404 `

I can't track down where the replicator would start to fetch these documents properly but once I can piece that together I can maybe get a better picture of what is missing.

**Comment by davidpfahler** [comment]
@robertjpayne I'd think that is the expected 404. Doesn't it log a message to the console saying "You can safely ignore this 404" or something like that?

**Comment by robertjpayne** [comment]
@davidpfahler no log message -- but yes agree that this 404 is expected, but for some reason this is where PouchDB then stops working to some degree.

**Comment by robertjpayne** [comment]
Well I've managed to track down one source of issues. Basically after the replication fails the app gets spammed with 3 warnings about `XMLHttpRequest: abort() cancels JS callbacks…`.

This is failing because of the request-browser.js dependency sets timers and the timers are not behaving quite like native browser timers are.

**Comment by robertjpayne** [comment]
@nolanlawson @davidpfahler so I've exhausted what I can to try and track this down. My best guess is this isn't related to the XMLHttpRequest or webpack but rather some other dependency that doesn't behave properly in react-native's environment.

* `XMLHttpRequest` is operating just fine as far as I can tell, it send the requests and proxies back the results. As far as I can see there is nothing it does to the request/response that would cause PouchDB to all of a sudden stop working.

* `request-browser.js` doesn't clean up it's timer that is set on line `162` for successful requests. This probably is minor but react-native throws a warning because you cannot abort in-flight XMLHttpRequests yet on the native side. Regardless this timer should be cleaned up.

* `replicate.js` starts doing it's work but `processPendingBatch` method around line `354` is the last thing that is called. 

I'd focus scoping in on the processPendingBatch method as it seems to be the last branch of code to work until it dies entirely.

**Comment by robertjpayne** [comment]
I should note I was mostly testing with log statements because the chrome debugger wasn't giving me any luck:

```js
if (pendingBatch.changes.length === 0) {
    // executes
      if (batches.length === 0 && !currentBatch) {
       // executes
        if ((continuous && changesOpts.live) || changesCompleted) {
          // executes
          returnValue.state = 'pending';
          returnValue.emit('paused');
          returnValue.emit('uptodate', result);
        }
        if (changesCompleted) {
          // does not execute
          completeReplication();
        }
      }
      // executes -- replication and PouchDB stop here
      return;
    }
```

**Comment by davidpfahler** [comment]
@robertjpayne thx for the input! Yes, in fact, I think you cannot use the chrome debugger because then the js executes in chrome and not in javascript core. However, what you can do is inspect the jscontext with safari. Do not enable safari debugging! That will execute the code in a webview, which we also don't want.

I'll look at your findings in detail when I can make time.

**Comment by robertjpayne** [comment]
@davidpfahler well got chrome debugger working -- you can execute it there and it persists with the same issue.

Unfortunately I don't really know what to look for but looking at all the variables and sorts it just looks like it thinks there are zero documents to sync from the server.

**Comment by nolanlawson** [comment]
It might be helpful to turn on PouchDB debug logging `PouchDB.debug.enable('*')` and then see if you can see any difference between running in Chrome (no errors) vs JSCore (errors).

In general, it sounds like there's either a bug in React Native's XHR implementation, or JSCore has some very weird XHR behavior that we're going to need to handle in `request-browser.js`. Oddly the PouchDB test suite passes just fine against PhantomJS, Safari 6+, and iOS 7+, so you'd think JSCore wouldn't be any different... But it does indeed seem to be a slightly different environment (e.g. the fact that `atob` and `btoa` are undefined).

**Comment by robertjpayne** [comment]
@nolanlawson I really doubt it's the XHR implementation, the network requests are running fine and the data coming back into PouchDB's JS is fine too.

My guess it's another rogue dependency.

**Comment by chandu0101** [comment]
any chance to resume this work again ...

**Comment by nolanlawson** [comment]
I do still think XHR is the problem (apparently React Native has a custom implementation, because JavaScriptCore does not have XHR), but nobody has tracked it down yet: https://github.com/nolanlawson/pouchdb-async-storage/issues/5

**Comment by abuisman** [comment]
What is the status of this now? Same as 28th of July?

**Comment by nolanlawson** [comment]
asyncstorage-down is no longer working in latest React for unknown reasons. Nobody is working on it right now: https://github.com/tradle/asyncstorage-down/issues/2

**Comment by nolanlawson** [comment]
I'm going to close this issue since it clearly should be solved by an adapter plugin, not in PouchDB core.

**Comment by andpor** [comment]
I am the maintainer of react-native-sqlite-storage. Would be happy to chip in to make it work with PouchDB...let me know. The interface is very similar to cordova-sqlite-storage.

**Comment by nolanlawson** [comment]
@andpor if you expose a `window.sqlitePlugin` variable (or if your users could do that), then the work is already done. :) https://github.com/pouchdb/pouchdb/blob/99c24a5861e19ac5ae78b35d71c9845a449ba9a0/src/adapters/websql/utils.js#L177

**Comment by andpor** [comment]
so there is no global window in ReactNative. I tried PouchDB on ReactNative and got a whole bunch of red screens with require errors etc so assume PouchDB is not compatible...

**Comment by abuisman** [comment]
@andpor Could you share/create a basic React-native project where you get these errors in? Then I/we could chip in in trying to figure out how to make it work.

**Comment by cem2ran** [comment]
stumbled into this: https://github.com/carbureted/react-native-pouchdb
Haven't tested it out yet.

**Comment by abuisman** [comment]
Why does PouchDB rely so much on `window.` ?

Also, wouldn't it be possible to pass another context for `window`? Pseudo: `(function(window){})({storage: sqlite})`

**Comment by andpor** [comment]
@abuisman I tried poc with this repo. Look at issue list...specifically this one..
https://github.com/nolanlawson/pouchdb-async-storage/issues/3

**Comment by vespakoen** [comment]
Working on supporting this **native**ly: https://github.com/facebook/node-haste/pull/46

**Comment by nwlieb** [comment]
For anyone that still wants to get this working, I have a way to get it very stable, based on the sqlite adapter method [here](https://github.com/carbureted/react-native-pouchdb) and the polyfills from [here](https://github.com/stockulus/pouchdb-react-native) along with my own additions. This works with pouchdb 5.4.5 and is heavily coupled to this version. I would not recommend upgrading if using this method.
```
// Polyfill for pouchdb works for PouchDB 5.4.5, DO NOT UPGRADE!
// Summary: pouch expects the following:
// -an implementation of the standard openDatabase which we fudge (new between 5.3.1->5.4.5)
// -the following polyfils: Buffer, atob, btoa, blob, process.nextTick
// -a websql plugin

global.Buffer = global.Buffer || require('buffer').Buffer;
global.atob = global.atob || require('atob');
global.btoa = global.btoa || require('btoa');
require('blob-polyfill');
window.openDatabase = true;
window.sqlitePlugin = require('react-native-sqlite-storage');
process.nextTick = process.nextTick || setImmediate;

const PouchDB = require('pouchdb-core');
PouchDB.plugin(require('pouchdb-adapter-websql'))
        .plugin(require('pouchdb-adapter-http'))
        .plugin(require('pouchdb-replication'))
        .plugin(require('pouchdb-upsert'));
```

Hope this helps someone, cheers.

**Comment by jareware** [comment]
Very nice, thanks for reporting back! Hoping this'll become easier with future versions.

**Comment by andpor** [comment]
I am using this package https://www.npmjs.com/package/react-native-pouchdb. Works fine.
Can these changes be incorporated into pouch tree?

**Comment by nolanlawson** [comment]
We have no intention of adding first-party React Native support. PouchDB is focused on the browser and Node.js, and the core of PouchDB is small enough that this problem can be solved in plugin-land.

Also there's more than one way to skin this cat - as @nwlieb points out, using SQLite via the SQLite plugin works fine as long as you have the right polyfills for RN's JSCore. But as @andpor's repo shows, you can also use AsyncStorage instead. Both of them have performance tradeoffs, but you can use either one with PouchDB and they should work fine.
